rust/lance-datafusion/src/spill.rs

```diff
@@ -0,0 +1,753 @@
+pub struct SpillReceiver {
+    status_receiver: tokio::sync::watch::Receiver<WriteStatus>,
+    path: PathBuf,
+    schema: Arc<Schema>,
+}
+
+impl SpillReceiver {
+    /// Returns a stream of batches from the spill. The stream will emit
+    /// batches as they are written to the spill. If the spill has already
+    /// been finished, the stream will emit all batches in the spill.
+    ///
+    /// The stream will not complete until [`Self::finish()`] is called.
+    ///
+    /// If the spill has been dropped, an error will be returned.
+    pub fn read(&self) -> SendableRecordBatchStream {
+        let rx = self.status_receiver.clone();
+        let reader = SpillReader::new(rx, self.path.clone());
+
+        let stream = futures::stream::try_unfold(reader, move |mut reader| async move {
+            match reader.read().await {
+                Ok(None) => Ok(None),
+                Ok(Some(batch)) => Ok(Some((batch, reader))),
+                Err(err) => Err(err),
+            }
+        });
+
+        Box::pin(RecordBatchStreamAdapter::new(self.schema.clone(), stream))
+    }
+}
+
+struct SpillReader {
+    pub batches_read: usize,
+    receiver: tokio::sync::watch::Receiver<WriteStatus>,
+    state: SpillReaderState,
+}
+
+enum SpillReaderState {
+    Buffered { spill_path: PathBuf },
+    Reader { reader: AsyncStreamReader },
+}
+
+impl SpillReader {
+    fn new(receiver: tokio::sync::watch::Receiver<WriteStatus>, spill_path: PathBuf) -> Self {
+        Self {
+            batches_read: 0,
+            receiver,
+            state: SpillReaderState::Buffered { spill_path },
+        }
+    }
+
+    async fn wait_for_more_data(&mut self) -> Result<Option<Arc<[RecordBatch]>>, DataFusionError> {
+        let status = self
+            .receiver
+            .wait_for(|status| {
+                status.error.is_some()
+                    || status.finished
+                    || status.batches_written() > self.batches_read
+            })
+            .await
+            .map_err(|_| {
+                DataFusionError::Execution(
+                    "Spill has been dropped before reader has finish.".into(),
+                )
+            })?;
+
+        if let Some(error) = &status.error {
+            let mut guard = error.lock().ok().expect_ok()?;
+            return Err(DataFusionError::from(&mut (*guard)));
+        }
+
+        if let DataLocation::Buffered { batches } = &status.data_location {
+            Ok(Some(batches.clone()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn get_reader(&mut self) -> Result<&AsyncStreamReader, ArrowError> {
+        if let SpillReaderState::Buffered { spill_path } = &self.state {
+            let reader = AsyncStreamReader::open(spill_path.clone()).await?;
+            // Skip batches we've already read.
+            for _ in 0..self.batches_read {
+                reader.read().await?;
+            }
+            self.state = SpillReaderState::Reader { reader };
```

Comment:
Ok...so if we've read batches 0, 1, 2, 3, 4...

And then the writer decides to spill.  It will still write batches 0, 1, 2, 3, 4?

This makes sense (as this will potentially need to be replayed multiple times) but it was confusing.

Reply:
I'll add a comment explaining that.